homeassistant/components/zha/climate.py

```diff
@@ -118,708 +42,124 @@ async def async_setup_entry(
 
     @property
     def fan_mode(self) -> str | None:
         """Return current FAN mode."""
-        if self._thrm.running_state is None:
-            return FAN_AUTO
-
-        if self._thrm.running_state & (
-            T.RunningState.Fan_State_On
-            | T.RunningState.Fan_2nd_Stage_On
-            | T.RunningState.Fan_3rd_Stage_On
-        ):
-            return FAN_ON
-        return FAN_AUTO
+        return self.entity_data.entity.fan_mode
 
     @property
     def fan_modes(self) -> list[str] | None:
         """Return supported FAN modes."""
-        if not self._fan:
-            return None
-        return [FAN_AUTO, FAN_ON]
+        return self.entity_data.entity.fan_modes
 
     @property
     def hvac_action(self) -> HVACAction | None:
         """Return the current HVAC action."""
-        if (
-            self._thrm.pi_heating_demand is None
-            and self._thrm.pi_cooling_demand is None
-        ):
-            return self._rm_rs_action
-        return self._pi_demand_action
-
-    @property
-    def _rm_rs_action(self) -> HVACAction | None:
-        """Return the current HVAC action based on running mode and running state."""
-
-        if (running_state := self._thrm.running_state) is None:
+        if self.entity_data.entity.hvac_action is None:
             return None
-        if running_state & (
-            T.RunningState.Heat_State_On | T.RunningState.Heat_2nd_Stage_On
-        ):
-            return HVACAction.HEATING
-        if running_state & (
-            T.RunningState.Cool_State_On | T.RunningState.Cool_2nd_Stage_On
-        ):
-            return HVACAction.COOLING
-        if running_state & (
-            T.RunningState.Fan_State_On
-            | T.RunningState.Fan_2nd_Stage_On
-            | T.RunningState.Fan_3rd_Stage_On
-        ):
-            return HVACAction.FAN
-        if running_state & T.RunningState.Idle:
-            return HVACAction.IDLE
-        if self.hvac_mode != HVACMode.OFF:
-            return HVACAction.IDLE
-        return HVACAction.OFF
-
-    @property
-    def _pi_demand_action(self) -> HVACAction | None:
-        """Return the current HVAC action based on pi_demands."""
-
-        heating_demand = self._thrm.pi_heating_demand
-        if heating_demand is not None and heating_demand > 0:
-            return HVACAction.HEATING
-        cooling_demand = self._thrm.pi_cooling_demand
-        if cooling_demand is not None and cooling_demand > 0:
-            return HVACAction.COOLING
-
-        if self.hvac_mode != HVACMode.OFF:
-            return HVACAction.IDLE
-        return HVACAction.OFF
+        return HVACAction(self.entity_data.entity.hvac_action)
 
     @property
     def hvac_mode(self) -> HVACMode | None:
         """Return HVAC operation mode."""
-        return SYSTEM_MODE_2_HVAC.get(self._thrm.system_mode)
+        if self.entity_data.entity.hvac_mode is None:
+            return None
+        return HVACMode(self.entity_data.entity.hvac_mode)
```

Comment:
Same as above

Reply:
added a mapping and updated to leverage HA internal property caching